### PR TITLE
DR-633 Exception restructuring

### PR DESCRIPTION
**Pull request by marikomedlock**
1. Made exceptions in common package abstract, except for NotImplemented. Changed places that were throwing one of these explicitly to throw a subclass instead.
2. Changed a few exception classes to extend a common one, instead of Runtime directly.
3. Added a comment to each common exception about the HTTP status code it maps to.
4. Only service/iam package knows about SAM ApiException. There it is converted to a DataRepoException before propagating outside.
    a. Added a conversion method in SamClientService to map SAM ApiExceptions to DataRepoExceptions based on their HTTP status code.
    b. Added try/catch/rethrow blocks to methods in the SAMClientService and removed them in its callers.
    c. Updated all places outside of the iam package that the SAM ApiException was being used/caught/declared and removed it. The only two exceptions are the global exception handler and the stubbed out SAM service used for the connected tests.
5. Only service/job package knows about StairwayException. There it is converted to a DataRepoException before propagating outside.
    a. Divided Stairway exceptions into two groups. Made both groups extend a Stairway-specific base class.
    b. RuntimeException subclasses should be prevented by the caller, instead of being handled after the fact. For example, police the arguments they're passing to stairway to avoid a null reference.
    c. Exception subclasses cannot be prevented by the caller, and so should be handled after the fact. For example, decide what to do when an internal, unrecoverable stairway error occurs.
    d. Converted Stairway exceptions to a new subclass of InternalServerErrorException inside JobService.

Doc is here:
https://docs.google.com/document/d/1m3LEyHaAC3H6vXfVyvxcFrKBmzkRLWwWUP16aq2DrPk/edit#

### CODE REVIEW

File: src/main/java/bio/terra/app/controller/RepositoryApiController.java [hunk 1]
```diff
@@ -4,27 +4,12 @@
 import bio.terra.app.utils.ControllerUtils;
 import bio.terra.controller.RepositoryApi;
 import bio.terra.app.controller.exception.ValidationException;
-import bio.terra.common.exception.InternalServerErrorException;
-import bio.terra.common.exception.UnauthorizedException;
+import bio.terra.model.*;
```
Comment (ddietterich) [inline src/main/java/bio/terra/app/controller/RepositoryApiController.java]: We don't use '*' imports. I can show you how to dissuade IntelliJ from generating them.

**Review by ddietterich — APPROVED** [review]

**Comment by ddietterich** [comment]
LGTM

File: src/main/java/bio/terra/app/controller/RepositoryApiController.java (hunk 1 above)
Comment (marikomedlock) [reply src/main/java/bio/terra/app/controller/RepositoryApiController.java]: Understood and thanks. Fixed in most recent commit.

**Review by thathert — APPROVED** [review]
LGTM!

**Pull request merged**